instrumentation/github.com/Shopify/sarama/otelsarama/producer_test.go

```diff
@@ -331,6 +335,171 @@ func TestWrapAsyncProducerError(t *testing.T) {
 	assert.Equal(t, "test", span.StatusMessage())
 }
 
+func TestWrapAsyncProducer_DrainsSuccessesAndErrorsChannels(t *testing.T) {
+	// Mock provider
+	sr := new(oteltest.SpanRecorder)
+	provider := oteltest.NewTracerProvider(oteltest.WithSpanRecorder(sr))
+
+	// Set producer with successes config and fill it with successes and errors
+	cfg := newSaramaConfig()
+	cfg.Producer.Return.Successes = true
+
+	mockAsyncProducer := mocks.NewAsyncProducer(t, cfg)
+	ap := WrapAsyncProducer(cfg, mockAsyncProducer, WithTracerProvider(provider))
+
+	wantInput := 5
+	for i := 0; i < wantInput; i++ {
+		mockAsyncProducer.ExpectInputAndSucceed()
+		ap.Input() <- &sarama.ProducerMessage{Topic: topic, Key: sarama.StringEncoder("foo2")}
+	}
+
+	wantErrros := 3
+	for i := 0; i < wantErrros; i++ {
+		mockAsyncProducer.ExpectInputAndFail(errors.New("test"))
+		ap.Input() <- &sarama.ProducerMessage{Topic: topic, Key: sarama.StringEncoder("foo2")}
+	}
+
+	ap.AsyncClose()
+
+	// Ensure it is possible to read Successes and Errors after AsyncClose
+	var wg sync.WaitGroup
+
+	gotInput := 0
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for range ap.Successes() {
+			gotInput++
+		}
+	}()
+
+	gotErrors := 0
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for range ap.Errors() {
+			gotErrors++
+		}
+	}()
+
+	wg.Wait()
```

Comment:
Not worth a change:
You should be able to do this synchronously correct?

Reply:
question if something like this is more readable:

```go

var gotSuccesses, gotErrors int
var successClosed, errorsClosed bool
for !successClosed && !errorsClosed  {
  select {
  case _, ok := <- ap.Successes():
    if !ok {
      successClosed = true
      continue
    }
    gotSuccesses++
  case _, ok := <- ap.Errors():
    if !ok {
      errorsClosed = true
      continue
    }
    gotErrors ++
  }
}
```